src/common/exec_utils.h

```diff
@@ -98,10 +98,13 @@ inline bool SetupDefaultBlobsOut(const std::vector<NDArray>& src,
     is_default = nd.IsDefaultData();
 #endif
     if (!is_default) {
-      NDArray temp = bufs != nullptr ? bufs->at(i) : NDArray(nd.shape(), nd.ctx(),
-                                                             true, nd.dtype());
 #if MXNET_USE_MKLDNN == 1
+        NDArray temp = bufs != nullptr ? bufs->at(i) : nd.IsMKLDNNData() ?
+            nd.Reorder2Default() : NDArray(nd.shape(), nd.ctx(), true, nd.dtype());
```

Comment:
when `MXNET_USE_MKLDNN == 1`, isn't `is_default` the same as `nd.IsMKLDNNData()`?
why do you still need to check `nd.IsMKLDNNData()` here?

Reply:
is_default could be false because it is a sparse ndarray. in which case we cannot call reorder2default and will just make a copy like we previously did. this does mean that kAddTo won't work as we are not preserving the data to tmp.

Reply:
the logic here looks very complex. i guess you might need to check `IsMKLDNNData` first before checking `bufs`.
Also, what happens if nd is a default dense array and req is kAddTo? should we use nd directly? Should you check kAddTo more explicitly?